crates/shared/src/price_estimation.rs

```diff
@@ -121,19 +123,58 @@ impl Estimate {
     }
 }
 
-#[async_trait::async_trait]
+pub type PriceEstimateResult = Result<Estimate, PriceEstimationError>;
+
 #[mockall::automock]
-pub trait PriceEstimating: Send + Sync {
-    async fn estimate(&self, query: &Query) -> Result<Estimate, PriceEstimationError> {
-        self.estimates(std::slice::from_ref(query))
-            .await
-            .into_iter()
-            .next()
-            .unwrap()
+pub trait PriceEstimating: Send + Sync + 'static {
+    // The '_ lifetime in the return value is the same as 'a but we need to write it as underscore
+    // because of a mockall limitation.
+
+    /// Returns one result for each query in arbitrary order. The usize is the index into the queries slice.
+    fn estimates<'a>(&'a self, queries: &'a [Query])
+        -> BoxStream<'_, (usize, PriceEstimateResult)>;
```

Comment:
Because we are eliding the returned lifetime, can you just:
```suggestion
    fn estimates(&self, queries: &[Query])
        -> BoxStream<'_, (usize, PriceEstimateResult)>;
```

Reply:
Looking at https://doc.rust-lang.org/reference/lifetime-elision.html it doesn't have the same meaning because the lifetime wouldn't be tied to the argument, only self.

```rust
fn args1<T: ToCStr>(&mut self, args: &[T]) -> &mut Command;                  // elided
fn args2<'a, 'b, T: ToCStr>(&'a mut self, args: &'b [T]) -> &'a mut Command; // expanded
```